src/hotspot/share/logging/logDecorators.hpp

```diff
@@ -68,24 +70,50 @@ class LogDecorators {
     Invalid
   };
 
+  class DefaultUndecoratedSelection {
+    friend class TestLogDecorators;
+    LogSelection _selection;
+
+    DefaultUndecoratedSelection() : _selection(LogSelection::Invalid) {}
+
+  public:
+    template<typename... Tags>
+    DefaultUndecoratedSelection(LogLevelType level, LogTagType first, Tags... rest) : _selection(LogSelection::Invalid) {
+      static_assert(1 + sizeof...(rest) <= LogTag::MaxTags + 1,
+                    "Too many tags specified!");
+
+      LogTagType tag_arr[LogTag::MaxTags + 1] = { first, rest... };
+
+      if (sizeof...(rest) == LogTag::MaxTags) {
+        assert(tag_arr[sizeof...(rest)] == LogTag::__NO_TAG,
+               "Too many tags specified! Can only have up to " SIZE_FORMAT " tags in a tag set.", LogTag::MaxTags);
+      }
+
+      _selection = LogSelection(tag_arr, false, level);
+    }
+
+    const LogSelection& selection() const { return _selection; }
+  };
```

Comment:
I am uncomfortable with this type erasure. `LogTagType[LogTag::MaxTags + 1 /* = 6 */]` -> `LogTagType*` -> `LogTagType[LogTag::MaxTags /* = 5 */]`. I think this should be rewritten so that `tag_arr` is typed as a `LogTagType[5]`. I think everywhere we have a `const LogTagType parameter[LogTag::MaxTags]` really should have been `const LogTagType (&parameter)[LogTag::MaxTags]` so that this would have been a compile error. 

My suggestion is to either do the following:
```suggestion
  public:
    DefaultUndecoratedSelection(LogLevelType level, LogTagType t0, LogTagType t1 = LogTag::__NO_TAG,
                                LogTagType t2 = LogTag::__NO_TAG, LogTagType t3 = LogTag::__NO_TAG,
                                LogTagType t4 = LogTag::__NO_TAG, LogTagType guard_tag = LogTag::__NO_TAG) : _selection(LogSelection::Invalid) {
      assert(guard_tag == LogTag::__NO_TAG, "Too many tags specified!");

      LogTagType tag_arr[LogTag::MaxTags] = { t0, t1, t2, t3, t4 };

      _selection = LogSelection(tag_arr, false, level);
    }

    const LogSelection& selection() const { return _selection; }
  };
```

or maybe even better, do what we do for the `LogTagSet` and have a static helper and a private constructor, so that we can turn all the asserts into compile errors. 

Something like:
```suggestion
    DefaultUndecoratedSelection(LogLevelType level, LogTagType t0, LogTagType t1, LogTagType t2,
                                LogTagType t3, LogTagType t4) : _selection(LogSelection::Invalid) {
      LogTagType tag_arr[LogTag::MaxTags] = { t0, t1, t2, t3, t4 };
      _selection = LogSelection(tag_arr, false, level);
    }
  public:

    template <LogLevelType Level, LogTagType T0, LogTagType T1 = LogTag::__NO_TAG, LogTagType T2 = LogTag::__NO_TAG,
              LogTagType T3 = LogTag::__NO_TAG, LogTagType T4 = LogTag::__NO_TAG,
              LogTagType GuardTag = LogTag::__NO_TAG>
    static DefaultUndecoratedSelection make() {
      STATIC_ASSERT(GuardTag == LogTag::__NO_TAG);
      return DefaultUndecoratedSelection(Level, T0, T1, T2, T3, T4);
    }

    const LogSelection& selection() const { return _selection; }
  };
```

And we can then use `LogDecorators::DefaultUndecoratedSelection::make<level, LOG_TAGS(...)>()` to create them.

Reply:
Thanks for your comments! I think having some static selection maker that calls a private constructor is the neatest choice here. I'll go for that